config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java

```diff
@@ -158,75 +158,97 @@ private void removeEntitiesForDeleteBundle(final List<Entity> entities) {
+                        loadPolicyDependencies(dependentPolicy, annotatedBundle, rawBundle, excludeReusable);
+                        break;
+                    case EntityTypes.ENCAPSULATED_ASSERTION_TYPE:
+                        Encass encass = rawBundle.getEncasses().get(dependency.getName());
+                        loadEncassDependencies(encass, annotatedBundle, rawBundle, excludeReusable);
+                        break;
+                    default:
+                        loadGatewayEntity(dependency, annotatedBundle, rawBundle);
                 }
             }
-            // Remove policies of excluded Encasses
-            excludedReusableEntities.forEach(ex -> {
-                Map<String, ? extends GatewayEntity> policyEntities =
-                        annotatedBundle.getEntities(entityTypeRegistry.getEntityClass(EntityTypes.POLICY_TYPE));
-                if (ex instanceof Encass) {
-                    removeExcludedReusablePolicy((Policy) policyEntities.get(((Encass) ex).getPolicy()), annotatedBundle);
-                } else if (ex instanceof Policy) {
-                    removeExcludedReusablePolicy((Policy) ex, annotatedBundle);
-                }
-            });
         }
     }
 
     /**
-     * Inserts the Gateway entity into the Entity Map. It skip adding the entity if excludeReusable is TRUE and the
-     * Gateway Entity is Reusable
+     * Loads the Encass and its dependencies
      *
-     * @param entityMapToUpdate Entity Map where GatewayEntity needs to be inserted
-     * @param key Gateway entity key to store against
-     * @param gatewayEntity Gateway entity to be inserted
-     * @param excludeReusable Exclude inserting reusable entity
+     * @param encass Encass policy for which gateway dependencies needs to be loaded.
+     * @param annotatedBundle Annotated Bundle for which bundle is being created.
+     * @param rawBundle Bundle containing all the entities of the gateway.
+     * @param excludeReusable Exclude loading Reusable entities as the dependencies of the policy
      */
-    private void insertGatewayEntity(Map entityMapToUpdate, String key, GatewayEntity gatewayEntity,
-                                     boolean excludeReusable, List<GatewayEntity> excludedReusableEntities) {
-        if (excludeReusable && gatewayEntity instanceof AnnotableEntity && ((AnnotableEntity) gatewayEntity).isReusable()) {
-            excludedReusableEntities.add(gatewayEntity);
-            return; // Return without inserting as its reusable entity
+    private void loadEncassDependencies(Encass encass, AnnotatedBundle annotatedBundle, Bundle rawBundle,
+                                        boolean excludeReusable) {
+        if (encass != null && !excludeGatewayEntity(Encass.class, encass, annotatedBundle, excludeReusable)) {
+            annotatedBundle.getEncasses().put(encass.getName(), encass);
```

Comment:
just a check, encass.getName() is giving you the name or null?

Reply:
@RaghuReddy `getName()` should not give null as we set it in `Encass.postLoad()`